src/adapter/src/coord.rs

```diff
@@ -1487,6 +1495,77 @@ impl Coordinator {
         Ok(())
     }
 
+    /// Returns an `as_of` suitable for bootstrapping the given index dataflow.
+    fn bootstrap_index_as_of(
+        &self,
+        cluster_id: ComputeInstanceId,
+        dataflow: &DataflowDescription<OptimizedMirRelationExpr>,
+    ) -> Antichain<Timestamp> {
+        // All inputs must be readable at the chosen `as_of`, so it must be at least the join of
+        // the `since`s of all dependencies.
+        let storage_ids = dataflow.source_imports.keys().copied().collect();
+        let compute_ids = dataflow.index_imports.keys().copied().collect();
+        let id_bundle = CollectionIdBundle {
+            storage_ids,
+            compute_ids: btreemap! {cluster_id => compute_ids},
+        };
+        let mut as_of = self.least_valid_read(&id_bundle);
```

Comment:
Some parts of the `bootstrap_*_as_of` functions are shared, which seems like a good opportunity to factor it out?

Reply:
I copied this code from `finalize_dataflow`, so there is certainly a refactoring opportunity here!